### PR TITLE
Extracting more data

### PR DESCRIPTION
**Pull request by giacomo-granzotto-mews**
Including relationships/security config.
Including filters (all pages/page level/visual container levels) + config (page/visual container levels) + DataTransforms (visual container) + Query (visual container)

## Proposed changes

Include more Schema data to output extraction folders.

Deserializing Layout to a concrete type so we can generate more outputs from the message.
Few fields had to be deserialized twice because are escaped so the result can be checked as JSON instead escaped string.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by jiri-vinarek — COMMENTED** [review]
Proposed changes look good! Can you also add unit tests for new extracted files? See https://github.com/jiri-vinarek/DataModelTools/tree/master/src

File: src/Extractor/Core/File.cs [hunk 1]
```diff
@@ -53,6 +56,31 @@ public static File FromExpression(Expression expression)
             return new File("expressions", fileName);
         }
 
+        public static File FromRelationship(Relationship relationship)
+        {
+            var fileName = Sanitize($"{relationship.Name}");
```
Comment (jiri-vinarek) [inline src/Extractor/Core/File.cs:61]: I think you can just call `Sanitize(relationship.Name)`.

File: src/Extractor/Core/File.cs [hunk 2]
```diff
@@ -53,6 +56,31 @@ public static File FromExpression(Expression expression)
...
+
+            return new File("relationships", fileName);
+        }
+
+        public static File FromRole(Role role)
+        {
+            var fileName = Sanitize($"{role.Name}");
```
Comment (jiri-vinarek) [inline src/Extractor/Core/File.cs:68]: I think you can just call `Sanitize(role.Name)`.

File: src/Extractor/Core/Extractor.cs [hunk 3]
```diff
@@ -11,23 +13,103 @@ public static class Extractor
...
+                .Concat(visualConfigs)
+                .Concat(visualQuery)
+                .Concat(visualDataTransforms);
+        }
+
+        public static Extract GetAllPagesFilter(Layout layout)
+        {
+            return new Extract(
+            new File("report", "AllPagesFilter.json"),
+                JsonConvert.SerializeObject(layout.Filters.ConvertToFilters(), Formatting.Indented)
```
Comment (jiri-vinarek) [inline src/Extractor/Core/Extractor.cs:53]: There is already `GetLayoutExtract` which saves whole `Layout` object into `layout.json` file. The `GetAllPagesFilter` will extract part just part of the `layout` and save it into `AllPagesFilter.json`. Therefore a change to the layout will generate 2 changed extract files. Is it necessary? Ideally, I'd keep just one extract file.

File: src/Extractor/Core/File.cs [hunk 4]
```diff
@@ -1,4 +1,7 @@
 ﻿using Extractor.Dto;
+using System.Collections.Concurrent;
+using System.ComponentModel;
+using System.Data;
```
Comment (jiri-vinarek) [inline src/Extractor/Core/File.cs:4]: These usings are not used - please remove them.

File: src/Extractor/Program.cs [hunk 5]
```diff
@@ -5,6 +5,7 @@
 using System.Text;
 using Newtonsoft.Json;
 using Extractor.Dto;
+using System.Text.RegularExpressions;
```
Comment (jiri-vinarek) [inline src/Extractor/Program.cs:8]: The using is not needed - please remove it.

File: src/Extractor/Core/File.cs (hunk 4 above)
Comment (giacomo-granzotto-mews) [reply src/Extractor/Core/File.cs]: Done

File: src/Extractor/Core/Extractor.cs (hunk 3 above)
Comment (giacomo-granzotto-mews) [reply src/Extractor/Core/Extractor.cs:51]: Done

File: src/Extractor/Core/Extractor.cs [hunk 6]
```diff
@@ -1,5 +1,7 @@
 ﻿using System;
+using System.Collections;
 using System.Collections.Generic;
+using System.Data;
```
[the remark this answers is not in the document]
Comment (giacomo-granzotto-mews) [reply src/Extractor/Core/Extractor.cs:5]: Done

File: src/Extractor/Core/Extractor.cs [hunk 7]
```diff
@@ -1,5 +1,7 @@
 ﻿using System;
+using System.Collections;
```
[the remark this answers is not in the document]
Comment (giacomo-granzotto-mews) [reply src/Extractor/Core/Extractor.cs]: Done

File: src/Extractor/Core/File.cs (hunk 2 above)
Comment (giacomo-granzotto-mews) [reply src/Extractor/Core/File.cs]: Done

File: src/Extractor/Core/File.cs (hunk 1 above)
Comment (giacomo-granzotto-mews) [reply src/Extractor/Core/File.cs]: Done

File: src/Extractor/Core/Extractor.cs [hunk 8]
```diff
@@ -11,23 +13,103 @@ public static class Extractor
...
+                .Concat(visualFilters)
+                .Concat(visualConfigs)
+                .Concat(visualQuery)
+                .Concat(visualDataTransforms);
+        }
+
+        public static Extract GetAllPagesFilter(Layout layout)
+        {
+            return new Extract(
+            new File("report", "AllPagesFilter.json"),
```
[the remark this answers is not in the document]
Comment (giacomo-granzotto-mews) [reply src/Extractor/Core/Extractor.cs:50]: Done

**Comment by giacomo-granzotto-mews** [comment]
@jiri-vinarek, apart from what you requested I made some changes to stop use some concrete types (changed to use dynamic( since we dont have any official documentation and we found some discrepancies between files.
So if PowerBI changes for some reason the extractor will be able track de difference anyway.

File: src/Extractor/Core/Extractor.cs [hunk 9]
```diff
@@ -11,23 +12,102 @@ public static class Extractor
...
-        public static Extract GetLayoutExtract(object layout)
+        public static Extract GetVisualDataTransforms(Section section, Visualcontainer container)
+        {
+            return new Extract(
+                File.FromVisualContainer(section, container, "DataTransforms"),
+                JsonConvert.SerializeObject(container.DataTransforms.ConvertToDataTransforms(), Formatting.Indented)
+            );
+        }
+
+        public static Extract GetLayoutExtract(Layout layout)
```
Comment (jiri-vinarek) [inline src/Extractor/Core/Extractor.cs:103]: `GetLayoutExtract` is now unused - please remove it.

File: src/Extractor.Tests/ExtractTests.cs [hunk 10]
```diff
@@ -124,8 +127,213 @@ private IEnumerable<Extract> GetExtracts(string fileName)
...
+        return Extractor.GetLayoutExtracts(layout);
+    }
+
     private string GetDataPath(string fileName)
     {
         return Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)!, "Data", fileName);
     }
+
+    [Test]
+    public void ExtractRelationShiptTest()
```
Comment (jiri-vinarek) [inline src/Extractor.Tests/ExtractTests.cs:143]: Typo - `ExtractRelationShiptTest` -> `ExtractRelationshipsTest`

File: src/Extractor/Core/Extractor.cs [hunk 11]
```diff
@@ -11,23 +12,102 @@ public static class Extractor
...
+                .Concat(visualFilters)
+                .Concat(visualConfigs)
+                .Concat(visualQuery)
+                .Concat(visualDataTransforms);
+        }
+
+        public static Extract GetAllPagesFilter(Layout layout)
+        {
+            return new Extract(
+            new File("report", "AllPagesFilter.json"),
```
Comment (jiri-vinarek) [inline src/Extractor/Core/Extractor.cs:50]: I'd indent it like this:
```
            return new Extract(
                new File("report", "AllPagesFilter.json"),
```

File: src/Extractor.Tests/ExtractTests.cs [hunk 12]
```diff
@@ -124,8 +127,213 @@ private IEnumerable<Extract> GetExtracts(string fileName)
...
+            )
+        };
+
+        extracts.Should().BeEquivalentTo(expectedExtracts);
+    }
+
+    [Test]
+    public void ExtractPageFiltersTest()
+    {
+        var extract = GetLayoutExtracts("PageFilter.json").FirstOrDefault(x => x?.File?.FileName == "Filters");
```
Comment (jiri-vinarek) [inline src/Extractor.Tests/ExtractTests.cs:250]: I get multiple warnings that the first `?` is not necessary:

image

image

**Review by jiri-vinarek — COMMENTED** [review]
Good job! Just a few minor questions/suggestions.

**Comment by giacomo-granzotto-mews** [comment]
@jiri-vinarek I've made the adjustments + small suggestions from Visual Studio (not so smart as Resharper but... :)

**Pull request merged**